src/query/v2/shard_request_manager.hpp

```diff
@@ -172,34 +177,34 @@ class ShardRequestManager : public ShardRequestManagerInterface {
   std::vector<VertexAccessor> Request(ExecutionState<ScanVerticesRequest> &state) override {
```

Comment:
We can simplify a lot. Plz go over the suggested changes and what I will write here. All the functions I mention (including the suggestions) are for ScanVertices but the similar changes/logic should apply to the rest of the Requests.

First, observe that `PollOnRequests` is almost identical with `AwaitOnFirstRequest` with a few differences: 
1. `Poll` returns a `bool` to indicate at least one of the requests is ready. However it suffers from a small issue that it always returns false even if there are available futures because `paginated_response_tracker.at(*shard_it) == PaginatedResponseState::PartiallyFinished` will always be true if the first request in the request list is `PartiallyFinished`. This complicates the logic of the loop later on. 
2. `AwaitOnRequest` blocks on the next pending request within the `paginated_response_tracker`. 

Now if you look carefully on the `Request` body: 

```
   // Outer loop to make sure we are handling the paginated responses as well.
    while (!state.shard_cache.empty()) {
      // 1. Send the requests.
      SendAllRequests(state);
```
First, the `while (!state.shard_cache.empty())` is a bug, because it won't be true for continuations that are not sent yet (that is, the second, third etc batch from a shard request). Second, we already called `SendAllRequests` which effectively scattered the requests to their corresponding shards. That means, that we can iteratively block with a few modifications on the `RsmClient` and the body of the `Request` (plz see explanations for each suggested change).